### PR TITLE
Generation of MPH's from python objects

### PR DESCRIPTION
**Pull request by GregBowyer**
@venkateshks @Zarkantho @andrewlook 
Thoughts ? 

I am going to push out for numpy arrays because thats simple, then cut a tag. I am currently wiring this version in the the back of spark code to get the MPH per job run.

### CODE REVIEW

File: cmph/__init__.py [hunk 1]
```diff
@@ -216,7 +242,76 @@ def _range_check(name, lower, value, upper=None):
...
+        hash_fns = [fn.lower() for fn in cfg.hash_fns]
+        if any(fn not in _HASH_FNS.keys() for fn in hash_fns):
+            raise ValueError("Invalid internal hash fn")
+
+        if not all(bool(fn) for fn in hash_fns):
+            raise ValueError("Invalid internal hash fn")
+    else:
+        hash_fns = []
+
+    _range_check('chd_keys_per_bin', 1, cfg.chd_keys_per_bin, 128)
```
Comment (jerluc) [inline cmph/__init__.py]: SANIITTYYY

File: cmph/_adapters.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    pySideAdapter = _PythonListAdapter(obj)
+
+    @ffi.callback('char*()')
+    def read_fn():
+        cstr = pySideAdapter.read()
+        if cstr is not None:
+            pySideAdapter.box = ffi.new('char[]', cstr)
+            return pySideAdapter.box
+        else:
+            return None
```
Comment (jerluc) [inline cmph/_adapters.py]: I'm assuming Python's FFI does a sane thing when it gets a `None` coming back?

File: cmph/_adapters.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        # CMPH rather irritatingly does not always check the pointer
+        # that comes back
+        if pySideAdapter.current is not None:
+            return ffi.sizeof('char') * (len(pySideAdapter.current))
+        else:
+            return 0
+
+    @ffi.callback('void()')
+    def rewind_fn():
+        return pySideAdapter.rewind()
```
Comment (jerluc) [inline cmph/_adapters.py]: Does this need to `return` even though it is `void`?

File: cmph/_adapters.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    @ffi.callback('void()')
+    def destroy_fn():
+        pass
+
+    # THIS IS VITAL - without this you are going to
+    # accidently GC the callbacks and freak C out
```
Comment (jerluc) [inline cmph/_adapters.py]: Good call :+1:

File: cmph/_adapters.py (hunk 2 above)
Comment (GregBowyer) [inline cmph/_adapters.py]: Yes

File: cmph/_adapters.py (hunk 3 above)
Comment (GregBowyer) [inline cmph/_adapters.py]: Its not harmful, but its a bit silly so I will remove it for clarity

File: cmph/pyadapter.c [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+void _cmph_io_function_adapter_rewind(void *data) {
+    cmph_io_func_adapter_t *func_bag = (cmph_io_func_adapter_t *)data;
+    func_bag->rewindfn();
+}
+
+void _cmph_io_function_adapter_dispose(void *data, char *key, cmph_uint32 keylen) {
+    // Currently the adapter code should ‟lose” this ref
+    // causing a GC of the box.
```
Comment (jerluc) [inline cmph/pyadapter.c]: Are there any `del`s that need to occur on the Python side to ensure this? Or is there some scoping that does it for you?

Comment (GregBowyer) [inline cmph/pyadapter.c]: They happen automatic when falling out of scope

**Pull request merged**